Add basic virtual env setup script.

todor-ivanov opened the pull request:
Fixes #10979

#### Status
IN TESTING

#### Description
With the current script a proper virtual environment is setup for the WMCore central services.  Together with satisfying all the package dependencies inside it based on the `requirements_py3.txt` coming with the WMCore code base. 

NOTE: This is still a work in progress and it is meant to also provide a fully configured and properly linked runtime environment .  


#### Is it backward compatible (if not, which system it affects?)
NO

#### Related PRs
None 

#### External dependencies / deployment changes
In general this should depend on the service configuration repository:
https://gitlab.cern.ch/cmsweb-k8s/services_config/

todor-ivanov committed:
Add basic virtual env setup script.

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+        # TODO: Here to probe for the underlying packaging system
+
+        sudo -l
+        sudo yum -y install python3 git zip unzip gcc python-devel mariadb mariadb-devel openssl openssl-devel
```
Why `yum`? I thought the script should be independent from Linux package manager? How this will work on debian distros? Do we really need this in a script? I suggest to remove it completely since it add additional function to check if proper set of libraries is present on a system.

vkuznet replied:
Moreover, I doubt that this should be enforced either. For instance, to developed or fix some code I may not need mariadb at all. It may only be required when I run specific service. You must identify bare minimum of libraries which are **REQUIRED** to run WMCore services, i.e. in other words without which packages WMCore service will not start. For instance I may not need zip, gcc, python-devel and even openssl to run python.

todor-ivanov replied:
I can remove it. This list of packages here is only the minimal set that I have accumulated throughout the multiple runs on a brand new VM created inside the CERN Openstack, which were breaking the setup. And yes you are correct we are aiming to manage mostly python based dependencies, so those  could be left outside this script and let the one who is about to deploy them to handle the missing pieces in advance.

todor-ivanov replied:
I have added the list of OS and non-python packages/libraries dependencies per WMCore service in the help message of the script. So as we discussed long ago those are not going to be on this script to resolve them. We are going to use some  other means to do so.

@vkuznet  I am Closing this conversation here as well.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+#!/bin/bash
+
+### Usage: Usage: deploy-centralvenv.sh -t <wmcore_tag> -c <central_services_url> [-v wmcore_path] [-p  <patches>] [-s  <service_names>] [-l  <component_list>]
```
This is just for consistency with the already presented batch of deployment scripts used before in the deployment area of WMCore 

https://github.com/dmwm/WMCore/blob/master/deploy/deploy-centralvm.sh  && 

https://github.com/dmwm/WMCore/blob/master/deploy/deploy-wmagent.sh

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+
+### Usage: Usage: deploy-centralvenv.sh -t <wmcore_tag> -c <central_services_url> [-v wmcore_path] [-p  <patches>] [-s  <service_names>] [-l  <component_list>]
+### Usage:
+### Usage:   -t  <wmcore_tag>       WMCore tag to be used for this deployment [Default: 2.0.0]
```
That's a good comment I was also thinking it would be useful to make it optional. I'll fix it.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+### Usage: Usage: deploy-centralvenv.sh -t <wmcore_tag> -c <central_services_url> [-v wmcore_path] [-p  <patches>] [-s  <service_names>] [-l  <component_list>]
+### Usage:
+### Usage:   -t  <wmcore_tag>       WMCore tag to be used for this deployment [Default: 2.0.0]
+### Usage:   -c  <central_services> Url to central services (e.g. tivanov-unit01.cern.ch)
```
Not sure about that  one. I was keeping it as a mandatory flag as a precaution from somebody not hitting an already deployed service by overlooking on what and where he is currently deploying.

todor-ivanov replied:
Coming back to this comment here. 
I am not sure how to proceed actually. Because, in reality, this variable is not used anywhere yet in the current script. Back then, when we were using the `cmsweb` scripts for deploying in a virtual environment, we were having the `BASE_URL` provided from the setup environment : [[1]](https://github.com/dmwm/deployment/blob/541068546b79b2646d021a3a7f81ddfe60f7e772/reqmgr2/config.py#L13), while we now have the link hard coded in the service configs depending on the configuration branch to be used [[2]](https://gitlab.cern.ch/cmsweb-k8s/services_config/-/blob/preprod/reqmgr2/config.py#L14). At the same time we would need to have the ability to setup outside those K8 clusters - for development and debugging or any other needs, so the `BASE_URL` will need to be configurable at deployment time (regardless of the config branch to be used). One possible approach would be a `sed` command over all the config files, but this I'd consider as really far from elegant solution. @vkuznet  @amaltaro any ideas  are very welcome. Thanks!
 
[1]
https://github.com/dmwm/deployment/blob/541068546b79b2646d021a3a7f81ddfe60f7e772/reqmgr2/config.py#L13 

[2]
https://gitlab.cern.ch/cmsweb-k8s/services_config/-/blob/preprod/reqmgr2/config.py#L14

vkuznet replied:
I think what you need to do is the following. Instead of central FE URL, you should allow user to specify configs to use. This approach is more flexible since I can have my local configs pointing to my local services, or I can have a config with mix URLs (some local and some central). So far you **force** users to use central service and I'm advising you to stop doing that since as developer I don't need central service and can run all things at place I want to (I can choose to use central FE or I can run local services).

todor-ivanov replied:
Well I have already answered this few comments bellow, but just to mention: the parameter is not mandatory any more, and the capability of providing user's configs is there.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+### Usage:   -v  <wmcore_path>      WMCore virtual environment target path to be used for this deployment [Default: ./WMCore.venv3]
+### Usage:   -p  <patches>          List of PR numbers
+### Usage:                          (in double quotes and space separated e.g. "5906 5934 5922")
+### Usage:   -s  <service_names>    List of service names to be patched
```
it is a legacy, coming from the time (actually still the case) when because of the RPM based deployment every service was having its own area and branch in the deployed WMCore tree. So when it comes to patching a live service it is every time a single patch needs to be applied on all the deployed components (or at least the set of components you know the patch is affecting). Hence the list of services to be patched. And here is a script which I personally use very often when patching a single component [1]. Of course we are talking here about getting rid of slowly getting rid of this structure and pushing it all to run from a single code base and then the list of services to be patched is going to be a redundant option for sure. I do like the idea of running and patching a single code base for both development and runtime. (Although just as a note here - we may have issues if we want multiple users to be running the different services (as it was with cmsweb before), unless we decide to solve that in the K8 deployment scripts - we still have the option to dedicate a set of pods per service so the privileges and authentication related issues may be addressed there). 

[1]
https://github.com/dmwm/WMCore/blob/master/bin/patchComponent.sh

todor-ivanov replied:
Just a note here @vkuznet . 
I did remove the list of services to be patched as you suggested and left only the list of patches to be applied, which is indeed much cleaner, but I still need to implement the actual call to the patching script. Yet to be done.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+### Usage:                          (in double quotes and space separated e.g. "5906 5934 5922")
+### Usage:   -s  <service_names>    List of service names to be patched
+### Usage:                          (in double quotes and space separated (e.g. "rqmgr2 reqmgr2ms")
+### Usage:   -l  <component_list>   List of components to be deployed [Default: "admin reqmgr2 reqmgr2ms workqueue reqmon acdcserver"]
```
The list is inherited from the VM deployment procedure so far. Still working on the details. Once I get to fighting the runtime  issues for the different components, I'll come back to refining this list.

todor-ivanov replied:
A `component` here is actually a specific service to be deployed (e.g. `reqmgr2`, `reqmon` or any of `reqmgr2ms-*` etc.) In general we should have a single `pypi` package created for each  one of them and a dedicated configuration in `service_configs`, but in reality few of them are aggregated under the same `pypi` package but with separate configs like the micro services. I have made the relevant notes inline and we are about to separate those properly in the future. 

As of the name, I agree `component` and `componentList` were not the best choices. These should be renamed to `service` and `serviceList`. I am going to change that.

todor-ivanov replied:
done

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+### Usage: Example: yes | ./deploy-centralvenv.sh -p "10003" -s reqmgr2ms -c tivanov-unit01.cern.ch
+### Usage:
+
+FULL_SCRIPT_PATH="$(realpath "${0}")"
```
Do you have any suggestions on what would work on them both? Or should we try to discover the underlying OS/Distribution somewhere from above. Which I can foresee a need for in at least two other places in the script. One being the underlying system packaging system for resolving the system dependency packages.

todor-ivanov replied:
Two alternatives of `_realPath` function has been suggested in the code. We can chose which one to use at any moment. I am marking this conversation as resolved here.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+usage()
+{
+    echo -e $1
+    /usr/bin/perl -ne '/^### Usage:/ && do { s/^### Usage:?//; print }' < $FULL_SCRIPT_PATH
```
That's a good remark. I'll get rid of `perl` here. Thanks!

todor-ivanov replied:
fixed

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+help()
+{
+    echo -e $1
+    /usr/bin/perl -ne '/^### Usage:/ && do { s/^### Usage:?//; print }' < $FULL_SCRIPT_PATH
```
noted

todor-ivanov replied:
done

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+venvPath="./WMCore.venv3"           # WMCore virtual environment target path
+venvPath=$(realpath $venvPath)
+wmSrcRelPath="WMCore"               # WMCore source code path relative to $venvPath
+wmTag="2.0.0"                       # wmcore tag
```
this here is basically just setting the default usually the tag is expected as a parameter to the script, but yes I can make it read the latest one similarly to what @amaltaro  suggested here [1]

[1]
https://github.com/dmwm/WMCore/pull/10911#discussion_r769550143

https://github.com/dmwm/WMCore/blob/master/src/python/WMCore/__init__.py#L9

todor-ivanov replied:
Done

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+serNameToPatch=""                   # a list of service Names to patch
+vmName=""                           # hostname for central services
+vmName=${vmName%%.*}
+pythonCmd=/usr/bin/python3
```
good comment. I simply never considered MaCOS before. The most I was having into mind were mostly the set of different linux based distributions. I'll fix that.

vkuznet replied:
it has nothing to do with MacOS. I may have linux and have my own python in my specific path. THerefore, it is issue with PATH env, that's it.

todor-ivanov replied:
fixed

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+    case ${opt} in
+        v)
+            venvPath=$OPTARG
+            venvPath=$(realpath $venvPath)
```
noted

vkuznet commented:
@todor-ivanov , the `realpath` can be easily replaced with this:
```
#!/bin/bash
FULL_SCRIPT_PATH="$(realpath "${0}")"
echo $FULL_SCRIPT_PATH
echo "$0"
```
both echo outputs produced the same results on Linux and MacOS if I run the full path of the script.

todor-ivanov commented:
yes, 
```
echo $0
```

Would definitely work for the Full script path but wont work here:
```
venvPath=$(realpath $venvPath)
```
I'll think of something.

todor-ivanov committed:
Add realPath && handleReturn functions.

todor-ivanov committed:
Add WMCore path to pythonLib && Change realPath return code 38.

todor-ivanov committed:
Remove Perl dependency && some more options

todor-ivanov commented:
Hi @vkuznet  I only need to create the proper runtime environment for the services. Which needs little bit of digging [[1]](https://github.com/dmwm/WMCore/issues/10979#issuecomment-1065167598). The rest is simply putting it here, which shouldn't be a problem. As we said this was not a planned development, and I am fighting to finalize what was in the the plans for Q1, so I work on this one only when I have some extra time, hence the issue is floating between waiting and work in progress. I will return to this soon.

todor-ivanov committed:
Expand deployent options && Add WMCore pkg install from pypi.

todor-ivanov committed:
Fix dependency resolution steps and error handling.

todor-ivanov committed:
Add virtual environment hooks for the first time.

todor-ivanov committed:
Add default paths structure for WMCore

todor-ivanov committed:
Adding WMCore virtual environment hooks.

todor-ivanov committed:
Setup WMCore env variables && Rename global scope functions && Reorder execution in main.

todor-ivanov committed:
Fix wrong return value due to inappropriate use of AND list.

todor-ivanov committed:
add extra pypi indexes to resolve dependency problems during deployment from test index.

todor-ivanov committed:
Better isolate WMCore deployment inside srv/ subtree.

todor-ivanov committed:
Porperly create current symlink and account for deployed version.

todor-ivanov committed:
Add noVenvCleanup option and the possiblity of reusing the space from previous installatins && Add verbouseMode option && Refactor options naming schema.

todor-ivanov committed:
Add --no-cache-dir option to pypi

todor-ivanov committed:
Build proper enabledList.

todor-ivanov committed:
Fix enabledList.

todor-ivanov committed:
Checkout service configs only for enabled services.

Typo

todor-ivanov committed:
Adding per service manage script creation.

todor-ivanov committed:
Fix missing mkdir for per-service path creation.

Typo

todor-ivanov committed:
Set security string at deployment time.

todor-ivanov committed:
Add the top level wmcmanage script.

todor-ivanov committed:
Add per service auth path into the PYTHONPATH && Fix top level wmcmanage script.

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,882 @@
+    cd $venvPath
+    # pip install $pipOpt wheel
+    # pip install $pipOpt --upgrade pip
+    pip install wheel
```
All pip commands should be run with `--user` and `--log` options to allow to install soft in user based location (and don't screw up with system area) and log everything it is doing which will allow to debug remote installation

todor-ivanov replied:
Hi @vkuznet sorry for not replying immediately to this comment, but I really wanted to test few things. So even though it sounds logical to have the ability to use `--user` with pip, it is conflicting with the whole virtual environment concept. Because the `--user` option does  mean to install all packages in the user defined area and location (defaulting to `~/.local/`), while virtual environment and even the current script  intends  to  completely isolate the setup inside a predefined destination. But never the less I did try it, and the result was as expected:
```
(WMCore.venv3) user1@:~/WMCore.venv3$ pip install --user user2 ipython
ERROR: Can not perform a '--user' install. User site-packages are not visible in this virtualenv.
```

So I'd suggest all user related procedures to be addressed from the caller side with `sudo` or alternative. 

I have similar suggestion for the `--log` option. Adding it to the `pip install` would cause only the python packages installations to be logged.  And we better have the whole output. I intentionally avoided dealing with `stdout` and `stderror` redirections from inside the script, because it makes it extremely uncomfortable to work with on a later stage. What I do usually when I want to log the output from the deployment process is simply tailing it to an external file like:
```
./deploy/deploy-centralvenv.sh -c cmsweb-testbed.cern.ch -n -y -s -i test -d /data/tmp/WMCore.venv3/ -j preprod -l wmcore==2.0.3.rc1 2>&1 |tee /data/tmp/deploy-centralvenv.log
```

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,882 @@
+    echo
+    echo "======================================================="
+    echo "Activate WMCore virtual env:"
+    echo -n "Continue? [y]: "
```
once again, make sure it can run without user interactions

todor-ivanov replied:
done

vkuznet reviewed (changes requested):
I suggest that you should list first all dependencies you rely on in this script, they should be shown in usage message and checked before execution of the script. Please document all functions. Please ensure that script can be run with and without user interaction.

todor-ivanov committed:
Find the proper OS python executable.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,309 @@
+    }
+}
+
+cleanVenv()
```
This function has been removed and replaced by a startup parameter.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,882 @@
+      -i <pypi-index>                The pypi index to use (i.e. prod or test) [Default: prod - pointing to https://pypi.org/simple/]
+      -h <help>                      Provides help to the current script
+
+    # Example: ./deploy-centralvenv.sh -c cmsweb-test1.cern.ch -i test -l wmcore==2.0.3rc1 -d /data/tmp/WMCore.venv3/ -m "Some security string"
```
Good point. I'll provide the extra info. 
Just to clarify: the current script is yet adapted only for central services deployment. We should have be another issue for addressing the `WMAgent` deployment inside a virtual environment.

todor-ivanov replied:
done

todor-ivanov committed:
Fix manage scripts.

todor-ivanov committed:
Rename componentList to serviceList.

todor-ivanov committed:
Update help message.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,882 @@
+    echo
+    echo "======================================================="
+    echo "Creating minimal virtual environment:"
+    echo -n "Continue? [y]: "
```
I am currently using the construction:
```
yes | ./deploy-centralvenv.sh [options...]
```
which does exactly the same but, yes I can add it as a parameter - much cleaner indeed.

todor-ivanov replied:
done

todor-ivanov committed:
Add assumeYes option.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,882 @@
+    echo
+    echo "======================================================="
+    echo "Cloning WMCore configuration files:"
+    echo -n "Continue? [y]: "
```
done

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,882 @@
+    echo "======================================================="
+    echo "Deployment parameters:"
+    echo "-------------------------------------------------------"
+    echo "componentList: $componentList"
```
done

todor-ivanov committed:
Add Function docstrings.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,882 @@
+    exit 1
+}
+
+_realPath(){
```
done

todor-ivanov committed:
Add checkNeeded finctioon for checking all current script dependencies.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,882 @@
+
+    [[ -d $wmSrcPath ]] ||  mkdir -p $wmSrcPath || return $?
+    cd $wmSrcPath
+    git clone $wmSrcRepo $wmSrcPath && git checkout $wmSrcBranch && [[ -n $wmTag ]] && git reset --hard $wmTag
```
done

todor-ivanov committed:
Add non-python DEPENDENCIES lists to the docstring.

todor-ivanov commented:
Hi @vkuznet  I think I've addressed all your comments and suggestions. Plenty of them I did find quite useful and the solutions I provided are doing the job so far. You let me know if the result is what you expected. Please take another look.
Thanks!

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,1000 @@
+    #               Here follows the list of all direct (first level) dependencies per service generated from those spec files:
+
+    #               acdcserver: [python3==3.8.2, rotatelogs==2.2.25, couchdb16==1.6.1]
+    #               reqmgr2ms : [python3==3.8.2, jemalloc==5.2.0, rotatelogs==2.2.25, mongo==3.2.10]
```
jemalloc is only required for optimization but it is not really required to run reqmgr2ms and other services. And, it should not be required since in dev mode I don't need it to run python server.

todor-ivanov replied:
Well for development purposes `jemalloc` may not be needed, but this is needed for the service to be run in reality. And If I understand correctly, at some point once we are ready to completly leave behind the RPM based setup, I  think we are about to use this script to deploy the services inside the container in order to have equivalent runtime and development environment. So these were the reasons to put it there. And also, without it, the so generated `manage` scripts per service are simply not usable. See line 771 in the current version:
https://github.com/dmwm/WMCore/blob/159835d434a34cb3a060d9387d8ce3c4f28917af/deploy/deploy-centralvenv.sh#L771

vkuznet replied:
No, I totally disagree, jemalloc is optional, I can successfully run `wmc-httpd` without it. What you need to do is to patch manage scripts to test jemalloc, if it exists you can use it, otherwise it should be empty string, e.g.
```
jemalloc=`command -v jemalloc`
...
$jemalloc wmc-httpd ...
```

todor-ivanov replied:
Well this is a nice trick. I like it.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,1000 @@
+    #          on the pameters set for configuring the deployment steps. This will avoid human intervention during deployment:
+    # ./deploy-centralvenv.sh -c cmsweb-test1.cern.ch -y -s -t 2.0.0.pre3 -p "10003 9998" -d /data/tmp/WMCore.venv3/ -m "Some security string"
+
+
```
You can deploy it on a laptop as it is right now. The FE to be used is still coming from the `service_config` files from gilab and for the time being this parameter plays no role here. We still need to take a decision if to keep it and apply it for the current deployment or to completely get rid of it and rely only on the parameters set at the `service_configs`. Please see my comment above on the topic: https://github.com/dmwm/WMCore/pull/10980#discussion_r859496766

vkuznet replied:
Here I asked to provide concrete example how to setup things on laptop w/o central service. As I wrote already in another comment I advise you to allow users to specify their configs. Doing this way I can easily provide my config and run things locally w/o any central service or FE.

todor-ivanov replied:
> Here I asked to provide concrete example how to setup things on laptop w/o central service

Every each of those examples I already provided are capable of executing on a laptop or where else a person would like to. They are not tied with anything, but enough available space at the destination. Anyway, I did provide an example and I have also gotten rid of the mandatory parameter as you asked.

When it comes to `w/o central service` I do not understand what you mean.... This script is the one deploying the central services. Maybe it is due to a misconception of terminology etc. but as it comes to WMCore and my understanding the FE is not the `Central services`  themselves. All services deployed from this script are central services. And as of now, because of the lack of separation between cmsweb deployment procedures and WMCore ones, simply the FE has always been part of central services as well, but it itself is not the whole of it. The rest is basically related to WMAgent components and external services (e.g. databases as a services etc.) which we use.

> Doing this way I can easily provide my config and run things locally w/o any central service or FE

You can already provide your won configs. Just upload them to a gitlab or any repository and point them through the options at deployment time like for the current default `-g https://:@gitlab.cern.ch:8443/cmsweb-k8s/services_config.git -j preprod` or what ever private repository and branch you wish.  Upon that once they are properly distributed at the services deployment area you may continue editing them as usual.

todor-ivanov committed:
Remove package versions from the DEPENDENCIES section of the script's docstring.

todor-ivanov commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,1000 @@
+    #               So far those has been resolved through the set of *.spec files maintained at: https://github.com/cms-sw/cmsdist/tree/comp_gcc630
+    #               Here follows the list of all direct (first level) dependencies per service generated from those spec files:
+
+    #               acdcserver: [python3==3.8.2, rotatelogs==2.2.25, couchdb16==1.6.1]
```
I have removed the package versions from the above list and rephrased the message so that people do not get the impression they should generate this list themselves.

todor-ivanov commented:
Thanks @vkuznet 
Please take a look at my replies above and the latest simple change I did based on your remarks. I hope this would be good enough.

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,1000 @@
+    exit 1
+}
+
+_realPath(){
```
Todor I still do not see docstring for this function, may be you missed to commit it.

todor-ivanov replied:
I just added it.

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,1000 @@
+verboseMode=false
+assumeYes=false
+noVenvCleanup=false
+secString=""
```
please document this variable

todor-ivanov replied:
done

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,1000 @@
+venvCleanup=true && $noVenvCleanup && venvCleanup=false
+
+# check for mandatory parameters:
+[[ -z $vmName ]] && usage "Missing mandatory argument: -c <central_services>"
```
please re-think this part and add ability to avoid mandatory central services. Instead, provide user configs.

todor-ivanov replied:
removed

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,1000 @@
+        local manageScript=${wmCfgPath}/${service}/manage
+        [[ -d ${wmCfgPath}/${service} ]] && touch $manageScript && chmod 755 $manageScript || { err=$?; echo "could not setup startup scripts for $service";  return $err  ;}
+        cat<<EOF>$manageScript
+#!/bin/bash
```
Can we avoid writing configuration file(s) from setup script? Can you commit this configuration file into some repo and use it? I think this approach is dangerous and will require lots of maintenance in a future. Instead all configs should be available through github/gitlab such that we can change them independently from this setup script.

todor-ivanov replied:
First, let me clarify something we are not writing config files here, but we are writing the `manage` scripts for the current services  deployed. This is exactly how it has been done from cmsweb deployment scripts for the `RPM` based deployment as  well. It is basically a pattern for a `manage` script provided here, and then it needs to be filled with a variable or two regarding the service at hand and then distributed at the right service area so that the global `wmcmanage` one can find every enabled service's `manage` script. I am talking about line [748-749](https://github.com/dmwm/WMCore/blob/9606ac57bd38ba3c6b758c4071b9f8a72ab27f8e/deploy/deploy-centralvenv.sh#L748-L749) and line [769](https://github.com/dmwm/WMCore/blob/9606ac57bd38ba3c6b758c4071b9f8a72ab27f8e/deploy/deploy-centralvenv.sh#L769) in the current version. And also the destination tied to the current deployment path on line [721](https://github.com/dmwm/WMCore/blob/9606ac57bd38ba3c6b758c4071b9f8a72ab27f8e/deploy/deploy-centralvenv.sh#L721). Those are variables defined at deployment time. If we can think of a proper way of providing them to the script (without using `sed`, which as already maentioned, I really find it far from elegant) those 'templates' can easily go to `WMCore/bin/.`  

Well for the `wmcmanage` one, we can already put it at `WMCore/bin`  as it is, because it does not depend on anything else but the list of `enabled` services. I have kept it here in order to have the whole process of creating all the `*manage` scripts at a single place.

vkuznet reviewed (changes requested):
@todor-ivanov you put a LOT of effort into this script, but you made it monolithic, i.e. it does everything. As such it has its pros/cons. For instance, as developer I can easily can do `git` actions, like checkout WMCore, and I don't know how it will coincide with this script. Can I point this script to use my git checkouts? On another hand, this is virtual env and it should contain everything we need. But you start embeding configs into this approach which I dislike and commented in a code. Please discard from script configs, make jemalloc optional, and provide ability to co-exists with existing `git` checkouts. I think the script will be much simpler if we'll allow users to do/use their own git actions. To give a reason about my preference to use own git actions is this. I have so many branches in my local git that if I'll use this venv script I'll need to port every single branch (which is not only extra work, extra space, etc. but it will force me to work only within venv area). What if I want to move to another node (CERN or FNAL) to continue my work. Then, I'll need to redo this again and again, i.e. I need to install venv, then I need to re-port my branches, etc. Please do not consider this as mandatory requirement, I'm trying to see how convenient will be such approach.

todor-ivanov committed:
Make FE as optional parameter && Relax jemalloc requirement && Add forgotten docstrings.

todor-ivanov committed:
Add extra protection for already deployed WMCore source.

todor-ivanov committed:
Adding extra protection for the cases when WMCore source points outside the current virtual env.

todor-ivanov commented:
Thanks @vkuznet  for making another review and all your comments. I think I have addressed more of them. Few of them with some new commits other by comments. If you do not mind, let me also try to shed some more light and clarify few things here as well. 

As a start let me express my view on the need of this script, which may help. We are in the process of phasing out the `RPM` based deployments and shifting more or less towards substituting this with another packaging/software distribution system, which as I can see should be `pypi`. (Need to stress - this process I do see as complementary to the containerization process rather than a substitution). This puts us in a very beneficial position in which we can use a well defined virtual environment with solidly set configuration and deployment tree inside it, with good isolation of our packages deployment area from the rest of the general purpose `pypi` deployed packages or OS dependent and general purpose libraries. This script is meant to provide also the sign of equality between the production runtime environment deployed inside a container running in production and the developers working environment. While at the same time does not tie developers' hands to not explore new concepts and ideas...., hence the ability to also run from source. All that said is simply leading to my vision about what our next step should be. Once we adopt it properly, and we are ready to get rid of the `RPM` based system we need to provide this script during the container building process with the proper set of options for the container at hand. That way we can have the same construction at runtime and during development, while at the same time the developers are not restricted to using a concrete version of the packages or configuration (!!! provided they take care of NOT messing up with the currently running production environment !!!! ).         

> you made it monolithic, i.e. it does everything. 

I tend to disagree here. None of the steps, besides the ones related to the virtual environment and deployment tree creation, are   mandatory and can easely be skiped during manual deployment.  

> For instance, as developer I can easily can do git actions, like checkout WMCore, and I don't know how it will coincide with this script. Can I point this script to use my git checkouts? 

Yes you can. I did put some extra protections with my last two commits to be 100% sure this script does not touch developer's working area. Upon initial deployment and configuration to run from source you can easily create a soft link at `$wmSrcPath` (by default it should be `${VIRTUAL_ENV}/srv/WMCore`) pointing to WMCore source code to somewhere else outside your virtual environment and upon that during any subsequent deployments you just provide the `-n` flag to the script so that you preserve your previous configuration and deployment area (and in case you forget to,  you are supposed to lose only the soft link)     

> On another hand, this is virtual env and it should contain everything we need. 

This I see it as supportive to what I have tried to achieve. 

> But you start embeding configs into this approach which I dislike and commented in a code. Please discard from script configs, 

This, as I already said to the comments inline, is:
 * First, not a mandatory step
 * Second, users config can be used instead of the mainstream configurations, provided they are properly uploaded in a repository accessible by `git` tools.


 > make jemalloc optional, 

I already did

> and provide ability to co-exists with existing git checkouts. I think the script will be much simpler if we'll allow users to do/use their own git actions. To give a reason about my preference to use own git actions is this. I have so many branches in my local git that if I'll use this venv script I'll need to port every single branch (which is not only extra work, extra space, etc. but it will force me to work only within venv area). 

The ability is already in place. I have added few extra protections on top of that, and mentioned that  as long as  the source code is mounted on the place where the deployment is happening it can be made accessible from inside the venv through a symlink.   

> What if I want to move to another node (CERN or FNAL) to continue my work. Then, I'll need to redo this again and again, i.e. I need to install venv, then I need to re-port my branches, etc. 

Again this is not the case. As long as your git working tree is physically accessible from the place where the deployment has taken place, it is going to work. 

On the other hand I can share a hint on what I am doing in my case (hope this can help someone). I am deploying somewhere remote (at any place where my central services are about to be able to access external services they depend on - like MongoDB etc.) and then I never login to the machine I have deployed at for code editing. Instead I work from my development environment on my local machine through emacs like:
```
user@localMachine $ emacs /ssh:lxplus.cern.ch:~/WMCoreDev.d/WMCore/src/python
```

While at the same time I do have this same source code linked from inside the virtual environment as:
```
user@VMorContainer $ ls -la /data/tmp/WMCore.venv3/srv/
d... auth
d... current -> /data/tmp/WMCore.venv3/srv/master
d... enabled
d... logs
d... master
d... state
l... WMCore -> ~/WMCoreDev.d/WMCore

```

This way I do not have delays from ssh lags for remote code development, while using the IDE from my own laptop.  And I do have my source wherever I need it, provided the area is physically mountable inside the container or VM.

vkuznet reviewed (commented):
@todor-ivanov instead of reviewing changes I tried to use it. Here is my findings using macOS:
```
./deploy-centralvenv.sh
./deploy-centralvenv.sh: line 219: md5sum: command not found
./deploy-centralvenv.sh: line 280: declare: -A: invalid option
declare: usage: declare [-afFirtx] [-p] [name[=value] ...]
=======================================================
Deployment parameters:
-------------------------------------------------------
serviceList          : wmcore
enabledList          :  reqmgr2 reqmgr2-tasks reqmgr2ms-transferor reqmgr2ms-monitor reqmgr2ms-output reqmgr2ms-rulecleaner reqmgr2ms-unmerged workqueue reqmon reqmon-tasks t0_reqmon t0_reqmon-tasks
venvPath             : ./WMCore.venv3
wmCfgRepo            : https://:@gitlab.cern.ch:8443/cmsweb-k8s/services_config.git
wmCfgBranch          : test
wmSrcRepo            : https://github.com/dmwm/WMCore.git
wmSrcBranch          : master
wmTag                :
serPatch             :
pypi Index           : prod
runFromSource        : false
Cleanup Virtual Env  : true
verboseMode          : false
assumeYes            : false
central services host:
secSring             :
=======================================================
Continue? [y]:
...
You still have 5 sec. to cancel before we proceed.

=======================================================
Creating minimal virtual environment:
Continue? [y]:
...

=======================================================
Activate WMCore virtual env:
Continue? [y]:
...
Collecting wheel
  Downloading wheel-0.37.1-py2.py3-none-any.whl (35 kB)
Installing collected packages: wheel
Successfully installed wheel-0.37.1
Requirement already satisfied: pip in ./lib/python3.9/site-packages (22.0.4)

=======================================================
Setup WMCore paths inside the virtual env:
Continue? [y]:
...
./deploy-centralvenv.sh: line 289: /Users/vk/tmp/WMCore/WMCore.venv3/bin:/opt/anaconda3/bin:/Users/vk/.cargo/bin:/opt/local/bin:/opt/local/sbin:/Users/vk/bin:/Users/vk/perl5/bin:/Users/vk/bin/OSX/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/opt/local/bin:/opt/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/opt/X11/bin:/Library/Apple/usr/bin:/Users/vk/Work/Languages/Go/gopath/bin:/opt/local/Library/Frameworks/Python.framework/Versions/Current/bin:/opt/local/share/java/hadoop-1.1.1/bin:/Users/vk/.fzf/bin: syntax error: operand expected (error token is "/Users/vk/tmp/WMCore/WMCore.venv3/bin:/opt/anaconda3/bin:/Users/vk/.cargo/bin:/opt/local/bin:/opt/local/sbin:/Users/vk/bin:/Users/vk/perl5/bin:/Users/vk/bin/OSX/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/opt/local/bin:/opt/local/sbin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/usr/local/go/bin:/opt/X11/bin:/Library/Apple/usr/bin:/Users/vk/Work/Languages/Go/gopath/bin:/opt/local/Library/Frameworks/Python.framework/Versions/Current/bin:/opt/local/share/java/hadoop-1.1.1/bin:/Users/vk/.fzf/bin")
```
Even though it fails with something here is my attempt to use it
```
# and now if I list my local area I see the following
ls
WMCore.venv3          deploy-centralvenv.sh

ls WMCore.venv3
WMCore.venv3 bin          include      lib          pyvenv.cfg

# and it seems to be non-active
type python
python is /opt/local/bin/python

# env | grep PYTHON returns nothing too
```
Since I know python venv I tried the following
```
ls WMCore.venv3/bin
Activate.ps1  activate.csh  pip           pip3.9        python3       wheel
activate      activate.fish pip3          python        python3.9

# let's activate it
source WMCore.venv3/bin/activate
```
and I see that `env | grep PYTHON` is still empty but python location has changed:
```
type python
python is /Users/vk/tmp/WMCore/WMCore.venv3/bin/python
```
and since there is not PYTHONPATH I can't load WMCore
```
python
Python 3.9.12 (main, Mar 25 2022, 04:56:27)
[Clang 13.0.0 (clang-1300.0.29.3)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import WMCore
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ModuleNotFoundError: No module named 'WMCore'
```
From this trial I think you should do the following:
- fix installation issues with `md5sum` and `declare`
- provide summary after installation of what is installed and where
- setup proper environment or print instructions how to do that
- provide dump of examples, and either abort if user didn't provide required flags or dump example of what user can do when installation is completed with specific flags.

todor-ivanov committed:
Add minimal bash version requirement && Add md5sum requirement && Make install packages step not mandatory.

todor-ivanov commented:
Thank you very much @vkuznet  for testing it. This is all indeed useful info.

> ix installation issues with md5sum and declare

The problem with the `declare -A` command is not from the script, but simply you are running a too old bash version - prior to 4.*, which makes the use of associative arrays impossible. And since this is a major bash functionality that the script relies on, the rest of your run was simply doomed to failure. 

This old bash version seems to be a common issue for OS X systems. See a proposed fix discussed at stackoverflow here[1](https://stackoverflow.com/questions/6047648/associative-arrays-error-declare-a-invalid-option), this [comment](https://stackoverflow.com/a/43948526) in particular explains exactly what one can expect for MacOS and a solution proposed in this [one](https://stackoverflow.com/a/17992349). I am adding a required minimal bash version for the script to run with my latest  commit.

As of `md5sum`, well this is my mistake - I forgot to add it to the list of required tools. Adding it now. The tool for `linux` based systems is provided I believe by `coreutils` package (naming may change depending on the distribution). 

> provide summary after installation of what is installed and where

It is already there. If your run was not to fail so dramatically you were supposed to see the following output as a summary of what has been set and where:

```
=======================================================
Setup the WMCore hooks at the virtual environment activate script
Continue? [y]: 
...
WMCoreVenvVars[WMCORE_SERVICE_ENABLED]	:	/data/tmp/WMCore.venv3/srv/enabled
WMCoreVenvVars[WMCORE_SERVICE_ROOT]	:	/data/tmp/WMCore.venv3/srv
WMCoreVenvVars[WMCORE_SERVICE_LOGS]	:	/data/tmp/WMCore.venv3/srv/logs
WMCoreVenvVars[PATH]	:	/data/tmp/WMCore.venv3/srv/WMCore/bin/:/data/tmp/WMCore.venv3/srv/current/bin/:/data/tmp/WMCore.venv3/bin:/afs/cern.ch/cms/caf/scripts:/afs/cern.ch/cms/common:/afs/cern.ch/cms/bin:/usr/sue/bin:/usr/local/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/opt/puppetlabs/bin:/afs/cern.ch/user/t/tivanov/bin
WMCoreVenvVars[WMCORE_SERVICE_CONFIG]	:	/data/tmp/WMCore.venv3/srv/current/config
WMCoreVenvVars[X509_USER_CERT]	:	/data/tmp/WMCore.venv3/srv/auth/dmwm-service-cert.pem
WMCoreVenvVars[WMCORE_SERVICE_STATE]	:	/data/tmp/WMCore.venv3/srv/state
WMCoreVenvVars[WMCORE_SERVICE_SRC]	:	/data/tmp/WMCore.venv3/srv/WMCore
WMCoreVenvVars[WMCORE_SERVICE_TMP]	:	/data/tmp/WMCore.venv3/srv/current/tmp
WMCoreVenvVars[PYTHONPATH]	:	/data/tmp/WMCore.venv3/srv/WMCore/src/python/:/data/tmp/WMCore.venv3/srv/current/auth/t0_reqmon-tasks:/data/tmp/WMCore.venv3/srv/current/auth/t0_reqmon:/data/tmp/WMCore.venv3/srv/current/auth/reqmon-tasks:/data/tmp/WMCore.venv3/srv/current/auth/reqmon:/data/tmp/WMCore.venv3/srv/current/auth/workqueue:/data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-unmerged:/data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-rulecleaner:/data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-output:/data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-monitor:/data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-transferor:/data/tmp/WMCore.venv3/srv/current/auth/reqmgr2-tasks:/data/tmp/WMCore.venv3/srv/current/auth/reqmgr2:/data/tmp/WMCore.venv3/srv/current/lib/python3.6/site-packages:/data/tmp/WMCore.venv3/lib/python3.6/site-packages
WMCoreVenvVars[WMCORE_SERVICE_AUTH]	:	/data/tmp/WMCore.venv3/srv/auth
WMCoreVenvVars[X509_USER_KEY]	:	/data/tmp/WMCore.venv3/srv/auth/dmwm-service-key.pem

``` 

I will work on making it more readable, though.

[Valentin](@vkuznet) Thanks for testing it. Please give it a try again and write back your findings. I am also interested to see what the outcome will be. Since I am not able to run it under MacOS I will rely on your feedback in order to fix issues related to porting it to MacOS.   

[1]
https://stackoverflow.com/questions/6047648/associative-arrays-error-declare-a-invalid-option

vkuznet commented:
@todor-ivanov , you are wrong about my shell:
```
echo $SHELL
/bin/zsh
```
So, take this into account. I'm using zsh and it is up to date (aligned with master HEAD :) Since your script uses `#!/bin/bash` it is also wrong since my bash (again which is up to date) is located at
```
type bash
bash is /opt/local/bin/bash
```
is also up to date:
```
bash --version
GNU bash, version 5.1.16(1)-release (x86_64-apple-darwin20.6.0)
```
So, don't blame me :)

vkuznet commented on deploy/deploy-centralvenv.sh:
```diff
@@ -0,0 +1,1045 @@
+#!/bin/bash
```
please fix it to
```
#!/usr/bin/env bash
```

todor-ivanov replied:
done

todor-ivanov committed:
Add a separate fucntion for alligned printing of all virtual env variables set

todor-ivanov committed:
Change shebang

todor-ivanov committed:
Add tree && _sort for the final printout.

todor-ivanov commented:
Hi @vkuznet  with the latest commit I did fix the printout of the final environment variables to be more human readable. Here is an example of what it looks like in one of my setups. I hope this is good enough. Please give it a try and let me know.

```
=======================================================
Printing the final WMCore virtual environment parameters and tree:
-------------------------------------------------------

/data/tmp/WMCore.venv3
├── bin
├── include
├── lib
│   └── python3.6
│       └── site-packages
├── lib64 -> lib
├── share
│   ├── doc
│   └── man
│       └── man1
└── srv
    ├── auth
    ├── current -> /data/tmp/WMCore.venv3/srv/master
    ├── enabled
    ├── logs
    │   ├── reqmgr2
    │   ├── reqmgr2ms-monitor
    │   ├── reqmgr2ms-output
    │   ├── reqmgr2ms-rulecleaner
    │   ├── reqmgr2ms-transferor
    │   ├── reqmgr2ms-unmerged
    │   ├── reqmgr2-tasks
    │   ├── reqmon
    │   ├── reqmon-tasks
    │   ├── t0_reqmon
    │   ├── t0_reqmon-tasks
    │   └── workqueue
    ├── master
    │   ├── auth
    │   ├── bin
    │   ├── config
    │   ├── couchapp
    │   ├── cx_Oracle-doc
    │   ├── data
    │   ├── etc
    │   ├── lib
    │   ├── lib64
    │   ├── man
    │   ├── share
    │   └── tmp
    ├── state
    │   ├── reqmgr2
    │   ├── reqmgr2ms-monitor
    │   ├── reqmgr2ms-output
    │   ├── reqmgr2ms-rulecleaner
    │   ├── reqmgr2ms-transferor
    │   ├── reqmgr2ms-unmerged
    │   ├── reqmgr2-tasks
    │   ├── reqmon
    │   ├── reqmon-tasks
    │   ├── t0_reqmon
    │   ├── t0_reqmon-tasks
    │   └── workqueue
    ├── WMCore -> /afs/cern.ch/user/.../WMCoreDev.d/WMCore
    └── WMCore.save
        ├── bin
        ├── deploy
        ├── doc
        ├── etc
        ├── src
        ├── standards
        ├── test
        └── tools

63 directories

-------------------------------------------------------
WMCoreVenvVars[PATH]                   : /data/tmp/WMCore.venv3/srv/WMCore/bin/
                                         /data/tmp/WMCore.venv3/srv/current/bin/
                                         /data/tmp/WMCore.venv3/bin
                                         /afs/cern.ch/cms/caf/scripts
                                         /afs/cern.ch/cms/common
                                         /afs/cern.ch/cms/bin
                                         /usr/sue/bin
                                         /usr/local/bin
                                         /usr/bin
                                         /usr/local/sbin
                                         /usr/sbin
                                         /opt/puppetlabs/bin
                                         /afs/cern.ch/user/.../bin
WMCoreVenvVars[PYTHONPATH]             : /data/tmp/WMCore.venv3/srv/WMCore/src/python/
                                         /data/tmp/WMCore.venv3/srv/current/auth/t0_reqmon-tasks
                                         /data/tmp/WMCore.venv3/srv/current/auth/t0_reqmon
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmon-tasks
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmon
                                         /data/tmp/WMCore.venv3/srv/current/auth/workqueue
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-unmerged
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-rulecleaner
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-output
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-monitor
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmgr2ms-transferor
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmgr2-tasks
                                         /data/tmp/WMCore.venv3/srv/current/auth/reqmgr2
                                         /data/tmp/WMCore.venv3/srv/current/lib/python3.6/site-packages
                                         /data/tmp/WMCore.venv3/lib/python3.6/site-packages
WMCoreVenvVars[RUCIO_HOME]             : /data/tmp/WMCore.venv3/srv/current
WMCoreVenvVars[WMCORE_SERVICE_AUTH]    : /data/tmp/WMCore.venv3/srv/auth
WMCoreVenvVars[WMCORE_SERVICE_CONFIG]  : /data/tmp/WMCore.venv3/srv/current/config
WMCoreVenvVars[WMCORE_SERVICE_ENABLED] : /data/tmp/WMCore.venv3/srv/enabled
WMCoreVenvVars[WMCORE_SERVICE_LOGS]    : /data/tmp/WMCore.venv3/srv/logs
WMCoreVenvVars[WMCORE_SERVICE_ROOT]    : /data/tmp/WMCore.venv3/srv
WMCoreVenvVars[WMCORE_SERVICE_SRC]     : /data/tmp/WMCore.venv3/srv/WMCore
WMCoreVenvVars[WMCORE_SERVICE_STATE]   : /data/tmp/WMCore.venv3/srv/state
WMCoreVenvVars[WMCORE_SERVICE_TMP]     : /data/tmp/WMCore.venv3/srv/current/tmp
WMCoreVenvVars[X509_USER_CERT]         : /data/tmp/WMCore.venv3/srv/auth/dmwm-service-cert.pem
WMCoreVenvVars[X509_USER_KEY]          : /data/tmp/WMCore.venv3/srv/auth/dmwm-service-key.pem

```

vkuznet reviewed (changes requested):
Todor, macOS does not have `md5sum` tool, therefore you need to adjust your script to manage md5 in a different way. Quick google [search](https://www.garron.me/en/bits/how-to-md5sum-mac-os-x.html) reveals that on maxOS we have `md5` tool which seems to me equivalent.

Bottom line, here is what I got:
```
./deploy-centralvenv.sh
./deploy-centralvenv.sh: line 219: md5sum: command not found
The current setup script requires: md5sum in order to continue. Please install it and rerun.
Interrupt execution due to step failure:
ERRORNO: 1
```

Now, regarding compatibility:
- On Linux we have
```
echo "Todor" | md5sum
91e53b1970dc62a271b70d24ab3c66e5  -
```
- on macOS we have
```
echo "Todor" | md5
91e53b1970dc62a271b70d24ab3c66e5
```

amaltaro reviewed (commented):
Todor, I finally managed to have a first look at this script, sorry for the delay.

It was hard to "parse" everything and closer to the end I didn't pay too much attention. Some general comments are:
* maybe we don't need to ask for input confirmation for actions that are not really changing something in someone's machine (like git clones).
* I think Valentin mentioned it in the past, but I do think it would be extremely helpful (and more sustainable) to have a small set of scripts. For instance, for CMSWEB, we have `prep`, `sw` and `post`, each in charge of different things. Here, we could easily have a script that does the initial virtual environment setup + python package requirements; separated from the actual WMCore services deployment; and maybe even separated from the WMCore service execution.
* I also see you integrated some service "manage" actions in this script, replacing the standard manage scripts we already have in the services_config/deployment repositories. Is this really necessary or is it worth it the extra complexity on our side?

todor-ivanov committed:
Fix missing _realPath call for default deployment destination.

todor-ivanov committed:
Add python executable as ext parameter && Fix package list && Resolve full pkg dependencies only when running from source.
